### PR TITLE
kanidm: make `serverSettings.db_path` writeable

### PR DESCRIPTION
**Pull request by TheRealGramdalf**
## Description of changes

Makes `services.kanidm.serverSettings.db_path` writeable to allow for custom database locations. This is in line with the [upstream example documentation](https://kanidm.github.io/kanidm/stable/server_configuration.html). 

## Things done

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) (or backporting [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) and [24.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

**Comment by TheRealGramdalf** [comment]
Should fix #323674 and allow a workaround for #312012

**Comment by TheRealGramdalf** [comment]
Maintainer CC: @erictapen @Flakebi

**Comment by TheRealGramdalf** [comment]
After looking into this further, this is more complicated than it seems. Additional work needs to be done with regards to `BindPaths` and `BindReadOnlyPaths`. At present the PR doesn't include the directory which houses the DB, which causes issues with the database lock.
Additionally, I found in my testing that adding the directory housing the database could fail if it coincided with a directory in `BindReadOnlyPaths` due (I believe) to the merge function [here](https://github.com/NixOS/nixpkgs/blob/ee8eb92f6be813b1ade668391cebd6e0123c058c/nixos/modules/services/security/kanidm.nix#L16)

TL;DR this is still a WIP, I aim to talk with upstream and sort this out when I've finished some other work.

**Review by steveej — APPROVED** [review]
merging this as-is makes sense to me to allow setting this option in the first place. currently, even with `lib.mkForce`, i'm getting the following error:

```
[ERROR]   stderr)        error: The option `containers.webserver.services.kanidm.serverSettings.db_path' is read-only, but it's set multiple times. Definition values:
[ERROR]   stderr)        - In `/nix/store/bjg8zxy01chf90zqn8jw8dq6nx8wnv83-source/nixos/modules/services/security/kanidm.nix': "/var/lib/kanidm/kanidm.db"
[ERROR]   stderr)        - In `/nix/store/a2gnsay8asfn2kqghz12nhw196cnmf8h-source/nixos/modules/virtualisation/nixos-containers.nix': "/var/lib/kanidm/db/kanidm.db"
```

this is when i'm trying to write expressions that deal with the read-only path issue -- which is different from the read-only nixos module option.
